core/ledger/kvledger/txmgmt/statedb/statecouchdb/statecouchdb.go

```diff
@@ -1098,6 +1098,9 @@ func (scanner *queryScanner) GetBookmarkAndClose() string {
 		retval = scanner.paginationInfo.bookmark
 	} else {
 		retval = scanner.queryDefinition.startKey
+		if retval == scanner.queryDefinition.endKey {
+			retval = ""
```

Comment:
Sorry, maybe I'm missing something but: we assign retval in line 1100 and then in line 1101 we test retval. How can the if ever be executed?

Reply:
The if will be executed when the scanner is exhausted, refer to the code [statecouchdb.go#L603](https://github.com/hyperledger/fabric/blob/1b887760ae6a0111fcf6bfdc18dd4e06173035d5/core/ledger/kvledger/txmgmt/statedb/statecouchdb/statecouchdb.go#L603); The nextStartKey is assigned to endKey when there is no more results, refer to [couchdb.go#L910](https://github.com/hyperledger/fabric/blob/1b887760ae6a0111fcf6bfdc18dd4e06173035d5/core/ledger/kvledger/txmgmt/statedb/statecouchdb/couchdb.go#L910).

The purpose of the if statement is to update the retval to match the behavior of leveldb when the scanner is exhausted.

Reply:
I see, thanks for the explanation. I would personally find the code a lot easier to read if you tested
```
scanner.queryDefinition.startKey == scanner.queryDefinition.endKey
```
but that's just a matter of taste